sydent/hs_federation/verifier.py

```diff
@@ -0,0 +1,206 @@
+            raise SignatureVerifyException("Signature missing")
+        for server_name, sigs in signed_json['signatures'].items():
+            if acceptable_server_names is not None:
+                if server_name not in acceptable_server_names:
+                    continue
+
+            server_keys = yield self._getKeysForServer(server_name)
+            for key_name, sig in sigs.items():
+                if key_name in server_keys:
+                    if 'key' not in server_keys[key_name]:
+                        logger.warn("Ignoring key %s with no 'key'")
+                        continue
+                    key_bytes = decode_base64(server_keys[key_name]['key'])
+                    verify_key = signedjson.key.decode_verify_key_bytes(key_name, key_bytes)
+                    logger.info("verifying sig from key %r", key_name)
+                    signedjson.sign.verify_signed_json(signed_json, server_name, verify_key)
+                    logger.info("Verified signature with key %s from %s", key_name, server_name)
+                    defer.returnValue((server_name, key_name))
+            logger.warn(
+                "No matching key found for signature block %r in server keys %r",
+                signed_json['signatures'], server_keys,
+            )
+        logger.warn(
+            "Unable to verify any signatures from block %r. Acceptable server names: %r",
+            signed_json['signatures'], acceptable_server_names,
+        )
+        raise SignatureVerifyException("No matching signature found")
+
+    @defer.inlineCallbacks
+    def authenticate_request(self, request, content):
+        """Authenticates a Matrix federation request based on the X-Matrix header
+        XXX: Copied largely from synapse
+        """
+        json_request = {
+            "method": request.method,
+            "uri": request.uri,
+            "destination_is": self.sydent.server_name,
+            "signatures": {},
+        }
+
+        if content is not None:
+            json_request["content"] = content
+
+        origin = None
+
+        def parse_auth_header(header_str):
+            try:
+                params = auth.split(" ")[1].split(",")
+                param_dict = dict(kv.split("=") for kv in params)
+
+                def strip_quotes(value):
+                    if value.startswith("\""):
+                        return value[1:-1]
+                    else:
+                        return value
+
+                origin = strip_quotes(param_dict["origin"])
+                key = strip_quotes(param_dict["key"])
+                sig = strip_quotes(param_dict["sig"])
+                return (origin, key, sig)
+            except Exception:
+                raise SignatureVerifyException("Malformed Authorization header")
+
+        auth_headers = request.requestHeaders.getRawHeaders(b"Authorization")
+
+        if not auth_headers:
+            raise NoAuthenticationError("Missing Authorization headers")
```

Comment:
what is the logic behind when we raise a SignatureVerifyException and when a NoAuthenticationError?

Reply:
Mostly just me thinking that the you might want separate errors for them to be slightly more helpful (eg. for if you spelt Authorization with an s or something). Granted, with the way we're using it here we could just have the same exception type with a different message.